lib/Feature/FeatureModelWriter.cpp

```diff
@@ -0,0 +1,342 @@
+    RC = xmlTextWriterWriteString(Writer, BAD_CAST(*F)->getName().data());
+    CHECK_RC
+    ++F;
+    for (; F != Group.end(); ++F) {
+      RC = xmlTextWriterWriteString(Writer, BAD_CAST " | ");
+      CHECK_RC
+      RC = xmlTextWriterWriteString(Writer, BAD_CAST(*F)->getName().data());
+      CHECK_RC
+    }
+
+    RC = xmlTextWriterEndElement(Writer); // CONSTRAINT
+    CHECK_RC
+  }
+
+  // TODO write other boolean constraints if they are parsed
+
+  RC = xmlTextWriterEndElement(Writer); // BOOLEANCONSTRAINT
+  return RC;
+}
+
+int FeatureModelXmlWriter::writeFeature(xmlTextWriterPtr Writer,
+                                        Feature &Feature1) {
+  int RC;
+
+  RC = xmlTextWriterStartElement(Writer, XmlConstants::CONFIGURATIONOPTION);
+  CHECK_RC
+
+  RC = xmlTextWriterWriteElement(Writer, XmlConstants::NAME,
+                                 BAD_CAST Feature1.getName().data());
+  CHECK_RC
+
+  // optional
+  RC = xmlTextWriterWriteElement(
+      Writer, XmlConstants::OPTIONAL,
+      BAD_CAST(Feature1.isOptional() ? "True" : "False"));
+  CHECK_RC
+
+  // parent
+  if (not Feature1.isRoot()) {
+    RC = xmlTextWriterWriteElement(
+        Writer, XmlConstants::PARENT,
+        BAD_CAST Feature1.getParent()->getName().data());
+    CHECK_RC
+  }
+
+  // children
+  if (Feature1.children_begin() != Feature1.children_end()) {
+    RC = xmlTextWriterStartElement(Writer, XmlConstants::CHILDREN);
+    CHECK_RC
+
+    OrderedFeatureVector Children;
+    Children.insert(Feature1.children());
+    for (Feature *F : Children) {
+      RC = xmlTextWriterWriteElement(Writer, XmlConstants::OPTIONS,
+                                     BAD_CAST F->getName().data());
+      CHECK_RC
+    }
+
+    RC = xmlTextWriterEndElement(Writer); // CHILDREN
+    CHECK_RC
+  }
+
+  // excludes
+  if (Feature1.excludes_begin() != Feature1.excludes_end()) {
+    RC = xmlTextWriterStartElement(Writer, XmlConstants::EXCLUDEDOPTIONS);
+    CHECK_RC
+
+    OrderedFeatureVector Exludes;
+    Exludes.insert(Feature1.excludes());
+    for (Feature *F : Exludes) {
+      RC = xmlTextWriterWriteElement(Writer, XmlConstants::OPTIONS,
+                                     BAD_CAST F->getName().data());
+      CHECK_RC
+    }
+
+    RC = xmlTextWriterEndElement(Writer); // EXCLUDEDOPTIONS
+    CHECK_RC
+  }
+
+  // implications
+  if (Feature1.implications_begin() != Feature1.implications_end()) {
+    RC = xmlTextWriterStartElement(Writer, XmlConstants::IMPLIEDOPTIONS);
+    CHECK_RC
+
+    OrderedFeatureVector Implies;
+    Implies.insert(Feature1.implications());
+    for (Feature *F : Implies) {
+      RC = xmlTextWriterWriteElement(Writer, XmlConstants::OPTIONS,
+                                     BAD_CAST F->getName().data());
+      CHECK_RC
+    }
+
+    RC = xmlTextWriterEndElement(Writer); // IMPLIEDOPTIONS
+    CHECK_RC
+  }
+
+  // if numeric
+  if (NumericFeature::classof(&Feature1)) {
+    auto &Nf = dynamic_cast<NumericFeature &>(Feature1);
```

Comment:
And if you directly need a num feature after wards use `llvm::dyn_cast<>` instead of `isa` in the if condition

```
if (NumericFeature *Nf = llvm:dyn_cast<NumericFeature>(Feature)) {
  ...
}
```